### DISCUSSION TITLE
Concepts Setup

### DISCUSSION THREAD

**Discussion by MentalGear** [sinclairzx81/typebox#766] [Q&A] [answered]
Typebox seems really optimized for performance, but for people coming from ZOD or other js validator the concepts can be a bit uncommon.

1. For Validating data, does one always have to compile the schema, or is there also a function to do it on the fly? For example, I have a class that uses the compiled schema for checking the input data during construction, but afterwards in the setters, uses 'Type.Pick' to check only against one property. 
2. If a compiling is necessary for each validation, is it possible to merge compiled validators into a new Type.Object in order to keep it DRY?
3. Is there an easy way to setup custom error messages (for [Superform](http://superforms.rocks)).

**Comment by sinclairzx81** [comment, marked as answer]
@MentalGear Hi

---

> For Validating data, does one always have to compile the schema, or is there also a function to do it on the fly? 

TypeBox supports both JIT and Dynamic checking. 

```typescript
import { TypeCompiler } from '@sinclair/typebox/compiler'
import { Value } from '@sinclair/typebox/value'
import { Type } from '@sinclair/typebox'

// JIT (fast)
const R1 = TypeCompiler.Compile(Type.String()).Check('hello')

// Dynamic (slow)
const R2 = Value.Check(Type.String(), 'hello')
```

> For example, I have a class that uses the compiled schema for checking the input data during construction, but afterwards in the setters, uses 'Type.Pick' to check only against one property.

Each type generated with TypeBox (including those derived from Pick) and considered distinct types. You will need to compile each separately (as it's not possible change logic to check only sub properties a type once it's compiled).

```typescript
import { TypeCompiler } from '@sinclair/typebox/compiler'
import { Type } from '@sinclair/typebox'

const Vector4 = Type.Object({
  x: Type.Number(),
  y: Type.Number(),
  z: Type.Number(),
  w: Type.Number(),
})
const Vector3 = Type.Pick(Vector4, ['x', 'y', 'z'])
const Vector2 = Type.Pick(Vector4, ['x', 'y'])
const Scalar  = Type.Index(Vector4, ['x'])

// compile each type separately
const C1 = TypeCompiler.Compile(Vector4)
const C2 = TypeCompiler.Compile(Vector3)
const C3 = TypeCompiler.Compile(Vector2)
const C4 = TypeCompiler.Compile(Scalar)
```
However, if you use `Value.Check`, you can just pass the type as dynamic checking does not require compilation.

---
> If a compiling is necessary for each validation, is it possible to merge compiled validators into a new Type.Object in order to keep it DRY?

Not currently. However there has been some consideration given to splitting JIT compiled modules into distinct functions that can be used to construct custom types. This isn't a feature of the library today (but may be explored in future)

---

> Is there an easy way to setup custom error messages (for [Superform](http://superforms.rocks/)).

You will need to refer to the documentation of superforms as different libraries handle error generation differently.

Hope this helps
S
